Implement JWT Authentication for regular users

Dragonfly911117 opened the pull request:
This PR primarily introduces JSON Web Token (JWT) authentication to protect the content within the FastAPI application. Here are the detailed changes:

1. JWT Authentication: Authentication mechanisms have been implemented to manage gathering/updating of accounts, shops, products, and image uploads. User permissions are enforced, ensuring authorized access and enhancing system security.

2. Misc: Improved error handling and file updates, including clearer error messages aligned with HTTP status code conventions. The README file now provides improved instructions for setting environment variables, aiding new contributors in project setup.

The primary focus of this PR is the implementation of JWT authentication for account/shop/product/image operations while improving code efficiency. These changes bolster the security, functionality, and maintainability of the FastAPI application.



## 中文好讀版:

這個 PR 主要是引入 JSON Web Token (JWT) 認證，以保護 FastAPI 應用程式中該被保護的東西。以下是變更的詳細內容：

1. JWT 認證: 已實施認證機制來管理帳戶、商店與商品的收集/更新以及圖片上傳。強制執行使用者許可權，確保授權訪問，提升系統安全性。

2. 其他增強功能: 改進了錯誤處理和檔案更新，包括與 HTTP 狀態碼慣例相符的更清晰的錯誤訊息。 README 檔案現在提供了改進的環境變數設定說明，幫助新貢獻者進行專案設定。

這個 PR 的主要焦點是實現帳號、商店、商品與圖片的 JWT 認證，同時改進了程式效率。這些變更增強了 FastAPI 應用程式的安全性、功能性和可維護性。

Dragonfly911117 committed:
Add JWT authentication

This commit adds JWT authentication support to the FastAPI application. More precisely, the python packages "passlib", "pydantic", and "python-jose[cryptography]" have been added to the requirements.txt. Other modifications include the addition of a new "JWT_test" directory with an "app.py" Python script for testing the JWT tokens along with the creation of "auth.py", "__init__.py", and "model/auth.py" in the existing "utils" and "model" folders. Also, an "access_token_expire_minutes" attribute has now been included in the core/settings.py file to define the duration of JWT tokens.

Dragonfly911117 committed:
Implementing JWT Authentication: This commit introduces JWT authentication support to the FastAPI application. Necessary packages are added to 'requirements.txt', while models and routers are updated to accommodate the changes. In addition, the token duration setting has been introduced in the main settings. An '.editorconfig' file is also added for maintaining consistent coding styles across various editors and IDEs.

Dragonfly911117 committed:
Refactor authentication to use JWT token: This commit introduces changes to the authentication system. The return value of 'authenticate_user' function is now TokenData instead of a string. This will make it easier to handle tokens on the front-end, as well as enhance the security by utilizing JWT tokens. In connection with this change, the import statement in '__init__.py' file was updated to include 'TokenData' class. Modifications were also made to 'login.py' and 'db_process.py' files to support the usage of the new TokenData type.

Dragonfly911117 committed:
Reduce login router prefix and refine authentication methods

The login router prefix has been reduced from `/login` to root for a cleaner URL. A significant refactor is also done to the authentication process, where JWT token is used to provide higher security level.

The `utils/auth.py` file now includes 'authenticate_user' function that returns the `TokenData` instead of a string. This change will make frontend development easier and will handle tokens more effectively. Two new functions 'get_account' and 'get_current_user' are added to return user profile and validate user authentication from token respectively.

As a result, the import statements and associated methods in `router/login.py` and `utils/__init__.py` have been updated to support the usage of the new `TokenData` type and authentication method.

Dragonfly911117 committed:
Streamlined Login Routes and Enhanced Authentication System
Simplified root login router for a cleaner URL. Revamped the authentication process for improved security through JWT tokens. Updated utils/auth.py so authenticate_user now returns TokenData for more effective token handling and simpler frontend development. Newly implemented methods get_account and get_current_user now return user profile and verify user authentication, respectively. Additionally, improvements were made for date formatting within the user account retrieval process. These collectively enhance security, user profile accuracy, and data handling efficiency.

Dragonfly911117 committed:
Refactor authentication and account management modules

The update is majorly on streamlining the login routes for clean URLs and enhancing the security of the authentication process. The auth.py utility file was revamped so that authenticate_user now returns TokenData instead of boolean for effective token handling. New methods for user account retrieval and user verification have also been integrated which respectively return user profile and confirm user authentication. The date formatting on the retrieval of user account details from the database was optimized. These changes collectively bolster security, enhance the accuracy of user data and improve the efficiency of data handling.

Dragonfly911117 committed:
Merge branch 'master' into auth

Dragonfly911117 committed:
Add instructions for managing environment variables

Included instructions on how to set environment variables using either .env file or IDE's run configurations in the README. Added this to help new contributors set up their development environments more easily and maintain an efficient working procedure, thereby enhancing project contribution rates, security, and data handling.

Dragonfly911117 committed:
Update response codes and error messages in account.py and db_process.py

Changed the HTTP response codes in account.py from 404 to 500, indicating that the error is on the server side, not that the resource was not found. Also, modified error messages to better reflect the nature of the error. Updated db_process.py to raise an HTTP exception instead of printing the error for better error handling and user feedback. These changes are made for better user experience and aligning with appropriate HTTP status code standards.

Dragonfly911117 committed:
Add authentication to image uploading and retrieval

Image.py has been updated to include authentication. The `upload_image` and `get_image` functions now depend on the active, authenticated user. This ensures that only users with appropriate access can upload or fetch images. The reference field on `if_exists_in_db` has also been changed from "shop_uuid" to "account_uuid" to appropriately validate the image's owner.

Dragonfly911117 committed:
Refactor image handling in API router

This commit significantly streamlines how images are handled in the API router. Previously, the logic related to image type validation and existence of owner in database was repeated. Now, separate functions have been created for these, reducing the complexity of the `upload_image` and `get_image` functions. Additionally, the `jsonable_encoder` has been removed as it was adding unnecessary complexity to JSONResponse, with dict proving to be sufficient. This refactor leads to improved readability and maintainability of the code.

Dragonfly911117 committed:
Refactor Shop routes and improve error handling

This commit refactor shop and image routes by improving error handling and response creation. The helper function `prepare_response` has been implemented to reduce redundancy in response creation, it will generate the proper response based on the result of the operations. A check is also added in `create_shop` route to assert that a shop does not already exist for a given account, this to prevent duplicate shops. The use of HTTPException has been introduced in image routes to provide detailed error messages when exceptional situations like non-existence of a shop for a given account. This refactor has improved the satiability and readability of the code.

Dragonfly911117 committed:
Add owner retrieval function in image handling

Imported Account in image_io.py and created a new owner retrieval function `get_owner_uuid`. This was done to allow us to flexibly determine the owner id based on the availability of shop or product uuid. Changes were applied in router/image.py to utilize the new function in both upload_image and get_image routes. This allows image handling to be specific for a shop or a product, improving the system's scalability and flexibility.

Dragonfly911117 committed:
Refactor image upload and retrieval code

Removed unused import and function `get_owner_uuid` from image_io.py. Replaced `is_image_available` with pre-existing checks in `if_exists_in_db` utility. Rewired `upload_image` and `get_image` functions in router/image.py to directly use these utilities for owner verification and retrieval. This eliminates redundancy, and makes the code cleaner and more maintainable. New functions `if_account_owns_shop` and `if_account_owns_product` were added in auth.py for detailed ownership validation. Function `if_one_owns_the_other` was added in db_process.py to verify ownership relationships.

Vghxv commented on utils/db_process.py:
```diff
@@ -51,8 +55,11 @@ def execute_sql(sql, param: Optional[Tuple] = None, fetch: bool = False) -> Unio
+    fetch = True
+    if sql.split()[0].upper() == "UPDATE" or sql.split()[0].upper() == "DELETE":
+        fetch = False
+    return execute_sql(sql, param, fetch=fetch)
```
programmer should know what to use for what sql operation, i think select sth and create/update operation do not have that much similarity.
btw, in function `execute_sql` actually do not need to connection.commit() when fetch == true(i.e. select) bcz database do not need update

Vghxv commented on utils/db_process.py:
```diff
@@ -86,3 +93,11 @@ async def if_exists_in_db(table_name, column_name, value) -> bool:
     assert result[0]["UUID_Exists"] in [0, 1]
     return result[0]["UUID_Exists"] == 1
+
+async def if_one_owns_the_other(
+        table_name: str, col1: str, val1: str, col2: str, val2: str
+) -> bool:
+    sql = f"SELECT EXISTS (SELECT 1 FROM {table_name} WHERE  {col1} = %s AND {col2} = %s) AS 'Exists';"
+    result = get_all_results(sql, (val1, val2))
+    assert result[0]["Exists"] in [0, 1]
+    return result[0]["Exists"] == 1
```
although i hate wrapper, i think this is acceptable bcz check database do confuse ppl a bit

Vghxv commented on router/shop.py:
```diff
@@ -1,128 +1,106 @@
             "model": ReturnShop
         },
-        status.HTTP_404_NOT_FOUND: {
+        404: {
```
any particular reason why use pure number?

PEKKA1117 replied:
forgot to import lib or the JBAI tried to fuck me, i guess.

Vghxv commented on router/shop.py:
```diff
@@ -1,128 +1,106 @@
+    content = jsonable_encoder({"msg": message, **data})
+
+    return JSONResponse(status_code=status_code, content=content)
+
```
function `create_shop` and `update_shop` , `if (result)` is evaluated inside and outside `prepare_response` which is kinda smelly.
More importantly, in function `get_shop`, `prepare_response` can return JSONResponse(status_code=status.HTTP_200_OK, content=content) when result is `false`

Vghxv reviewed (changes requested):
> What does a 500 Internal Server Error mean?
> 
> A 500 Internal Server Error occurs when a server is unable to fulfill a request sent to it due to an unexpected blocker. The server then sends that error message to indicate there’s an issue with it but it can’t seem to diagnose the problem.
in our Backend `server`, i don't think  if statements is a `unexpected blocker`

Dragonfly911117 committed:
"Refactor response handling and error messages in router and utils

Streamlined the response handling process in router/shop.py by removing the redundant prepare_response function and directly returning JSONResponse. This makes the code more readable and straightforward.

In utils/db_process.py, simplified the return handling in get_all_results and execute_query functions by removing redundant 'if' conditions and commit calls, thus improving code readability. Revised error messages to make them more user-friendly. These changes aim to enhance code maintainability and user experience."

Dragonfly911117 committed:
drive

Vghxv reviewed (approved):
lgtm
